lib/utils/matrix_sdk_extensions/download_file_extension.dart

```diff
@@ -0,0 +1,157 @@
+import 'dart:io';
+
+import 'package:fluffychat/data/network/media/media_api.dart';
+import 'package:fluffychat/di/global/get_it_initializer.dart';
+import 'package:fluffychat/utils/matrix_sdk_extensions/event_extension.dart';
+import 'package:matrix/matrix.dart';
+import 'package:path_provider/path_provider.dart';
+
+extension DownloadFileExtension on Event {
+  bool canContainAttachment() {
+    return [EventTypes.Message, EventTypes.Sticker].contains(type);
+  }
+
+  bool isSending() {
+    return status.isSending;
+  }
+
+  Uri? getAttachmentOrThumbnailMxcUrl(bool getThumbnail) {
+    return attachmentOrThumbnailMxcUrl(getThumbnail: getThumbnail);
+  }
+
+  bool isEncryptionDisabled(bool isEncrypted) {
+    return isEncrypted && !room.client.encryptionEnabled;
+  }
+
+  Future<FileInfo?> downloadOrRetrieveAttachment(
+    Uri mxcUrl,
+    String savePath,
+  ) async {
+    final database = room.client.database;
+    final attachment = await database?.getFileEntity(mxcUrl);
+
+    final mediaApi = getIt.get<MediaAPI>();
+    final downloadLink = mxcUrl.getDownloadLink(room.client);
+
+    if (attachment != null) {
+      return FileInfo(
+        filename,
+        attachment.path,
+        content.tryGet<int>('size') ?? attachment.lengthSync(),
+      );
+    }
+
+    final downloadResponse = await mediaApi.downloadFileInfo(
+      uriPath: downloadLink,
+      savePath: savePath,
+    );
+    if (downloadResponse.statusCode == 200) {
+      return FileInfo(
+        filename,
+        savePath,
+        content.tryGet<int>('size') ?? File(savePath).lengthSync(),
+        progressCallback: downloadResponse.onReceiveProgress,
+      );
+    }
+    throw ('getFileInfo: Download file $filename failed');
+  }
+
+  // Decrypt the file if it's encrypted.
+  Future<FileInfo?> decryptFile(
+    FileInfo? fileInfo,
+    Uri mxcUrl,
+    String decryptedPath, {
+    getThumbnail = false,
+  }) async {
+    final encryptedService = EncryptedService();
+    final fileMap = getThumbnail ? infoMap['thumbnail_file'] : content['file'];
+    if (!fileMap['key']['key_ops'].contains('decrypt')) {
+      throw ("getFileInfo: Missing 'decrypt' in 'key_ops'.");
+    }
+
+    final encryptedFile = EncryptedFileInfo.fromJson(fileMap);
+
+    if (!File(decryptedPath).existsSync()) {
```

Comment:
existsSync ?
it can block UI thread ?

Reply:
you right, lol the chat GPT shit suggests me :)